chainer/datasets/sub_dataset.py

```diff
@@ -0,0 +1,157 @@
+import numpy
+import six
+
+from chainer.dataset import dataset_mixin
+
+
+class SubDataset(dataset_mixin.DatasetMixin):
+
+    """Subset of a base dataset.
+
+    SubDataset defines a subset of a given base dataset. The subset is defined
+    as an interval of indexes, optionally with a given permutation.
+
+    If ``order`` is given, then the ``i``-th example of this dataset is the
+    ``order[start + i]``-th example of the base dataset, where ``i`` is a
+    non-negative integer. If ``order`` is not given, then the ``i``-th example
+    of this dataset is the ``start + i``-th example of the base dataset.
+    Negative indexing is also allowed: in this case, the term ``start + i`` is
+    replaced by ``finish + i``.
+
+    Args:
+        dataset: Base dataset.
+        start (int): The first index in the interval.
+        finish (int): The next-to-the-last index in the interval.
+        order (sequence of ints): Permutation of indexes in the base dataset.
+            If this is None, then the ascending order of indexes is used.
+
+    """
+    def __init__(self, dataset, start, finish, order=None):
+        if start < 0 or finish > len(dataset):
+            raise ValueError('subset overruns the base dataset.')
+        self._dataset = dataset
+        self._start = start
+        self._finish = finish
+        self._size = finish - start
+        self._order = order
```

Comment:
Sorry, I missed `DatasetMixin` raises NotImplementError in `__len__`. Then, implicit check is OK.